File: src/shady/generator_constructors.c
#include "generator.h"

static void generate_pre_construction_validation(Growy* g, json_object* src) {
    json_object* nodes = json_object_object_get(src, "nodes");
    shd_growy_append_formatted(g, "void pre_construction_validation(IrArena* arena, Node* node) {\n");
    shd_growy_append_formatted(g, "\tswitch (node->tag) { \n");
    assert(json_object_get_type(nodes) == json_type_array);
    for (size_t i = 0; i < json_object_array_length(nodes); i++) {
        json_object* node = json_object_array_get_idx(nodes, i);
        String name = json_object_get_string(json_object_object_get(node, "name"));
        String snake_name = json_object_get_string(json_object_object_get(node, "snake_name"));
        void* alloc = NULL;
        if (!snake_name) {
            snake_name = to_snake_case(name);
            alloc = (void*) snake_name;
        }
        shd_growy_append_formatted(g, "\tcase %s_TAG: {\n", name);
        json_object* ops = json_object_object_get(node, "ops");
        if (ops) {
            assert(json_object_get_type(ops) == json_type_array);
            for (size_t j = 0; j < json_object_array_length(ops); j++) {
                json_object* op = json_object_array_get_idx(ops, j);
                String op_name = json_object_get_string(json_object_object_get(op, "name"));
                String class = json_object_get_string(json_object_object_get(op, "class"));
                if (!class)
                    continue;
                bool list = json_object_get_boolean(json_object_object_get(op, "list"));
                if (strcmp(class, "string") == 0) {
                    if (!list)
                        shd_growy_append_formatted(g, "\t\tnode->payload.%s.%s = string(arena, node->payload.%s.%s);\n", snake_name, op_name, snake_name, op_name);
                    else
                        shd_growy_append_formatted(g, "\t\tnode->payload.%s.%s = _shd_import_strings(arena, node->payload.%s.%s);\n", snake_name, op_name, snake_name, op_name);
                } else {
                    String cap = capitalize(class);
                    shd_growy_append_formatted(g, "\t\t{\n");
                    String extra = "";
                    if (list) {
                        shd_growy_append_formatted(g, "\t\t\tsize_t ops_count = node->payload.%s.%s.count;\n", snake_name, op_name);
                        shd_growy_append_formatted(g, "\t\t\tLARRAY(const Node*, ops, ops_count);\n");
                        shd_growy_append_formatted(g, "\t\t\tif (ops_count > 0) memcpy(ops, node->payload.%s.%s.nodes, sizeof(const Node*) * ops_count);\n", snake_name, op_name);
                        shd_growy_append_formatted(g, "\t\t\tfor (size_t i = 0; i < ops_count; i++) {\n");
                        shd_growy_append_formatted(g, "\t\t\tconst Node** pop = &ops[i];\n");
                        extra = "\t";
                    }
                    if (!list)
                        shd_growy_append_formatted(g, "\t\t\tconst Node** pop = &node->payload.%s.%s;\n", snake_name, op_name);

                    shd_growy_append_formatted(g, "\t\t\t*pop = _shd_fold_node_operand(%s_TAG, Nc%s, \"%s\", *pop);\n", name, cap, op_name);

                    if (!(json_object_get_boolean(json_object_object_get(op, "nullable")) || json_object_get_boolean(json_object_object_get(op, "ignore")))) {
                        shd_growy_append_formatted(g, "%s\t\t\tif (!*pop) {\n", extra);
                        shd_growy_append_formatted(g, "%s\t\t\t\tshd_error(\"operand '%s' of node '%s' cannot be null\");\n", extra, op_name, name);
                        shd_growy_append_formatted(g, "%s\t\t\t}\n", extra);
                    }

                    shd_growy_append_formatted(g, "%s\t\t\tif (arena->config.check_op_classes && *pop != NULL && !is_%s(*pop)) {\n", extra, class);
                    shd_growy_append_formatted(g, "%s\t\t\t\tshd_error_print(\"Invalid '%s' operand for node '%s', expected a %s\");\n", extra, op_name, name, class);
                    shd_growy_append_formatted(g, "%s\t\t\t\tshd_error_die();\n", extra);
                    shd_growy_append_formatted(g, "%s\t\t\t}\n", extra);
                    if (list) {
                        shd_growy_append_formatted(g, "\t\t\t}\n");
                        shd_growy_append_formatted(g, "\t\t\tnode->payload.%s.%s = shd_nodes(arena, ops_count, ops);\n", snake_name, op_name);
                    }
                    free((void*) cap);
                    shd_growy_append_formatted(g, "\t\t}\n");
                }
            }
        }
        shd_growy_append_formatted(g, "\t\tbreak;\n");
        shd_growy_append_formatted(g, "\t}\n", name);
        if (alloc)
            free(alloc);
    }
    shd_growy_append_formatted(g, "\t\tdefault: break;\n");
    shd_growy_append_formatted(g, "\t}\n");
    shd_growy_append_formatted(g, "}\n\n");
}

void generate(Growy* g, json_object* src) {
    generate_header(g, src);

    json_object* nodes = json_object_object_get(src, "nodes");
    generate_pre_construction_validation(g, src);
}

File: src/shady/generator_constructors.py
from src.generator.generator import *
from src.generator.generator_common import *
from src.generator.generator_main import main

def generate_pre_construction_validation(g, src):
    nodes = src["nodes"]
    g.g += "void pre_construction_validation(IrArena* arena, Node* node) {\n"
    g.g += "\tswitch (node->tag) { \n"
    assert type(nodes) is list
    for node in nodes:
        name = node["name"]
        snake_name = node["snake_name"]

        g.g += f"\tcase {name}_TAG: {{\n"
        ops = node.get("ops")
        if ops is not None:
            assert type(ops) == list
            for op in ops:
                op_name = op["name"]
                clazz = op.get("class")
                if clazz is None:
                    continue
                is_list = op.get("list")
                if clazz == "string":
                    if is_list != True:
                        g.g += f"\t\tnode->payload.{snake_name}.{op_name} = shd_string(arena, node->payload.{snake_name}.{op_name});\n"
                    else:
                        g.g += f"\t\tnode->payload.{snake_name}.{op_name} = _shd_import_strings(arena, node->payload.{snake_name}.{op_name});\n"
                else:
                    classless = clazz == "none"
                    cap = capitalize(clazz)
                    g.g += "\t\t{\n"
                    extra = ""
                    if is_list == True:
                        g.g += f"\t\t\tsize_t ops_count = node->payload.{snake_name}.{op_name}.count;\n"
                        g.g += f"\t\t\tLARRAY(const Node*, ops, ops_count);\n"
                        g.g += f"\t\t\tif (ops_count > 0) memcpy(ops, node->payload.{snake_name}.{op_name}.nodes, sizeof(const Node*) * ops_count);\n"
                        g.g += f"\t\t\tfor (size_t i = 0; i < ops_count; i++) {{\n"
                        g.g += f"\t\t\tconst Node** pop = &ops[i];\n"
                        extra = "\t"
                    else:
                        g.g += f"\t\t\tconst Node** pop = &node->payload.{snake_name}.{op_name};\n"

                    if classless:
                        g.g += f"\t\t\t*pop = _shd_fold_node_operand({name}_TAG, 0, \"{op_name}\", *pop);\n"
                    else:
                        g.g += f"\t\t\t*pop = _shd_fold_node_operand({name}_TAG, Nc{cap}, \"{op_name}\", *pop);\n"

                    if not (op.get("nullable") or op.get("ignore")):
                        g.g += f"{extra}\t\t\tif (!*pop) {{\n"
                        g.g += f"{extra}\t\t\t\tshd_error(\"operand '{op_name}' of node '{name}' cannot be null\");\n"
                        g.g += f"{extra}\t\t\t}}\n"

                    if not classless:
                        g.g += f"{extra}\t\t\tif (arena->config.check_op_classes && *pop != NULL && !is_{clazz}(*pop)) {{\n"
                        g.g += f"{extra}\t\t\t\tshd_error_print(\"Invalid '{op_name}' operand for node '{name}', expected a {clazz}\");\n"
                        g.g += f"{extra}\t\t\t\tshd_error_die();\n"
                        g.g += f"{extra}\t\t\t}}\n"

                    if is_list == True:
                        g.g += "\t\t\t}\n"
                        g.g += f"\t\t\tnode->payload.{snake_name}.{op_name} = shd_nodes(arena, ops_count, ops);\n"

                    g.g += "\t\t}\n"
        g.g += "\t\tbreak;\n"
        g.g += "\t}\n"

    g.g += "\t\tdefault: break;\n"
    g.g += "\t}\n"
    g.g += "}\n\n"

def generate(g, src):
    generate_header(g, src)

    generate_pre_construction_validation(g, src)

main(generate)